agent/engine/containermetadata/acquire_metadata.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/cihub/seelog"
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+// acquireNetworkMetadata parses the docker.NetworkSettings struct and
+// packages the desired metadata for JSON marshaling
+func acquireNetworkMetadata(settings *docker.NetworkSettings) (NetworkMetadata, error) {
+	var err error
+	if settings == nil {
+		err = fmt.Errorf("Failed to acquire network metadata: Network metadata not available or does not exist")
+		return NetworkMetadata{}, err
+	}
+
+	// Get Port bindings from docker settings
+	ports, err := api.PortBindingFromDockerPortBinding(settings.Ports)
+	if err != nil {
+		ports = nil
+	}
+
+	// This metadata is available in two different places in NetworkSettings
+	gateway := settings.Gateway
+	iPAddress := settings.IPAddress
+	iPv6Gateway := settings.IPv6Gateway
+
+	// Assume there is at most one network mode (And if no mode, network is "none")
+	networkModeFromContainer := ""
+	if len(settings.Networks) == 1 {
+		for modeFromSettings, containerNetwork := range settings.Networks {
+			networkModeFromContainer = modeFromSettings
+			gateway = containerNetwork.Gateway
+			iPAddress = containerNetwork.IPAddress
+			iPv6Gateway = containerNetwork.IPv6Gateway
+		}
+	} else if len(settings.Networks) == 0 {
+		networkModeFromContainer = "none"
+	}
+	networkMD := NetworkMetadata{
+		ports:       ports,
+		networkMode: networkModeFromContainer,
+		gateway:     gateway,
+		iPAddress:   iPAddress,
+		iPv6Gateway: iPv6Gateway,
+	}
+	return networkMD, err
+}
+
+// acquireDockerContainerMetadata parses the metadata in a docker container
+// and packages this data for JSON marshaling
+func acquireDockerContainerMetadata(container *docker.Container) (DockerContainerMD, error) {
+	var err error
+	if container == nil {
+		err = fmt.Errorf("Failed to acquire container metadata: Container metadata not available or does not exist")
+		return DockerContainerMD{}, err
+	}
+
+	imageNameFromConfig := ""
```

Comment:
Early on I was encountering issues where the containers did not have a Config, probably due to testing on mocked containers. This check is for any metadata that comes from the Docker Container configuration, from only image name is kept currently. A possible case is if some problem occurs in the docker inspect call and we get a nil container config, which will throw a nil pointer exception if we do not handle it here.